src/util/range.h

```diff
@@ -0,0 +1,348 @@
+  {
+    return &(*current.get());
+  }
+
+  explicit map_iteratort(
+    iteratort underlying,
+    iteratort underlying_end,
+    std::shared_ptr<
+      std::function<value_type(const typename iteratort::value_type &)>> f)
+    : underlying(std::move(underlying)),
+      underlying_end(std::move(underlying_end)),
+      f(std::move(f))
+  {
+    if(this->underlying != this->underlying_end)
+      current = util_make_unique<outputt>((*this->f)(*this->underlying));
+  }
+
+  map_iteratort(const map_iteratort &other)
+    : underlying(other.underlying),
+      underlying_end(other.underlying_end),
+      f(other.f)
+  {
+    if(other.current != nullptr)
+      current = util_make_unique<outputt>(*other.current.get());
+  }
+
+private:
+  iteratort underlying;
+  iteratort underlying_end;
+  std::shared_ptr<
+    std::function<value_type(const typename iteratort::value_type &)>>
+    f;
+
+  /// Stores the result of \c f at the current position of the iterator.
+  /// Equals nullptr if the iterator reached \c underlying_end.
+  std::unique_ptr<outputt> current = nullptr;
+};
+
+/// Iterator which only stops at element for which some given function \c f is
+/// true.
+template <typename iteratort>
+class filter_iteratort
+{
+public:
+  using difference_type = void;
+  using value_type = typename iteratort::value_type;
+  using pointer = const value_type *;
+  using reference = const value_type &;
+  using iterator_category = std::forward_iterator_tag;
+
+  bool operator==(const filter_iteratort &other) const
+  {
+    return underlying == other.underlying;
+  }
+
+  bool operator!=(const filter_iteratort &other) const
+  {
+    return !(this->underlying == other.underlying);
+  }
+
+  /// Preincrement operator
+  filter_iteratort &operator++()
+  {
+    ++underlying;
+    point_to_first_to_peek();
+    return *this;
+  }
+
+  /// Postincrement operator
+  const filter_iteratort operator++(int)
+  {
+    filter_iteratort tmp(*this);
+    this->operator++();
+    return tmp;
+  }
+
+  const value_type &operator*() const
+  {
+    return *underlying;
+  }
+
+  const value_type *operator->() const
+  {
+    return &(*underlying);
+  }
+
+  filter_iteratort(
+    std::shared_ptr<std::function<bool(const value_type &)>> f,
+    iteratort underlying,
+    iteratort end)
+    : underlying(std::move(underlying)),
+      underlying_end(std::move(end)),
+      f(std::move(f))
+  {
+    point_to_first_to_peek();
+  }
+
+private:
+  iteratort underlying;
+  const iteratort underlying_end;
+  std::shared_ptr<std::function<bool(const value_type &)>> f;
+
+  /// Ensure that the underlying iterator is always positioned on an element
+  /// for which `f` is true.
+  /// This does nothing if \c f is satisfied at the current position.
+  void point_to_first_to_peek()
+  {
+    while(underlying != underlying_end && !(*f)(*underlying))
+      ++underlying;
+  }
+};
+
+/// Iterate other a first iterator and when the end is reached, iterate other
```

Comment:
s/other/over/ - and you don't iterate over an iterator. I think this needs to be reworded completely.